### PR TITLE
Test coverage for ensuring the MessageBodyWriter is used with null Accept header

### PR DESCRIPTION
**Pull request by jcarranzan**
### Summary

Coverage for backports 3.8.6 --> https://github.com/quarkusio/quarkus/pull/41411 - _Ensure that MessageBodyWriter is passed the proper media type._
Verify that currently , it works with `quarkus-rest` 

Notice this was failing with quarkus version:
` mvn clean verify -Dquarkus.platform.version=3.11.2 -Dit.test=HeadersIT#testWithNoAcceptHeader -Dreruns=0`

Also, I noticed that the fix is not working with `quarkus-resteasy`,I'll open a bug related to.

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [x] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [ ] Methods and classes used in PR scenarios are meaningful
- [ ] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java:111]: You don't need `CoreMatchers.is` the `is` is enough as we import `org.hamcrest.Matchers.is`

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java:104]: Can't we use `Header` type insted of `final var`?

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:109]: Shouldn't we use new endpoint like you did in `http-advanced-reactive`?

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:100]: Can you add tag `@Tag("https://github.com/quarkusio/quarkus/pull/41411")`

Also I think `HeadersMessageBodyWriter` should work with both or at least I would create issue if it's not work.

As reading this test it's work with that specific header but not with `MediaType.TEXT_PLAIN`. So I would say it's a bug

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java
Comment (jedla97) [inline http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: I think this can be `String`

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersMessageBodyWriter.java
Comment (jedla97) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersMessageBodyWriter.java:25]: I think this can be string

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:105]: Can't we use `Header` type insted of final var?

**Review by jedla97 — CHANGES_REQUESTED** [review]
Thanks for coverage. I have some comments and suggestion.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java
Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: what is wrong with `var`?

Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: I mean, it exists to cut verbosity/unnecessary info where the type is obvious (AFAICT, no idea what is official reason), and here, you know :-)

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java
Comment (michalvavrik) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java:46]: +1 for removal

Comment (michalvavrik) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java:46]: Logging is always useful when in trouble, but I don't fancy paying price (logging slows down, longer logs) "just for case". If this test become flaky or has hard to reproduce failure, we can add it.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java
Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: though I don't think @jcarranzan cares about that :-D it seems like this is more or less c&p for https://github.com/quarkusio/quarkus/pull/41411/files :-/

**Review by michalvavrik — CHANGES_REQUESTED** [review]
What is the point of coverage like this?

Upstream has completely same Quarkus REST tests, don't you trust upstream CI that has to pass or do you expect something specific for a product? I don't think we should mirror every test in upstream. @fedinskiy do we really need to test this? Cannot you use upstream test coverage label?

**Comment by michalvavrik** [comment]
>Test coverage for backport 3.8.6-41411
Please improve PR title, it will take just a few seconds to make clear what 41411 is. Thank you

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java:46]: Doing this test coverage I realized that line don't give any value and it's not needed as point Michal.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java
Comment (jedla97) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: > what is wrong with var?

I would say nothing just that I would use that in more complicated assigment of variable (like concatenating multiple strings together) or in foreach where I don't care that much about type. 

Anyway I'll leave the decision to @jcarranzan what he want to use.

Comment (jcarranzan) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: IMHO, var or String here it's irrelevant, at least if readability get worse, I would use strong typed as String.
Well @michalvavrik u are wrong in this case, I didn't realize about that code  and upstream would be a c&p from https://github.com/jholleran/message-body-writer-selection and so on...

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java
Comment (michalvavrik) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java:46]: Cool, thanks.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java
Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: > I would say nothing just that I would use that in more complicated assigment of variable (like concatenating multiple strings together) or in foreach where I don't care that much about type.

Ok, I don't care, I was just curious as I use it all the time.

Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: > Well @michalvavrik u are wrong in this case, I didn't realize about that code and upstream would be a c&p from https://github.com/jholleran/message-body-writer-selection and so on...

got it

Comment (michalvavrik) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersMessageBodyWriter.java:25]: btw

> concatenating multiple strings together

=> always string so `String` doesn't provide me a new info :-)

**Comment by michalvavrik** [comment]
> Maybe, I am missing something in the backports testing process, What is your way of proceeding in these cases?
Thanks

Yeah @jcarranzan , your work is alright. The reason why I am asking @fedinskiy (responsible for this release) is that I would simply replace this with JIRA label. Because there are some backports that is alright not to test. So I cannot let what's next, all my comments are rather opinions, you don't need to follow them :-)

**Comment by michalvavrik** [comment]
Anyway @jcarranzan @jedla97 feel free to proceed with this PR, I don't mean to obstruct.

**Comment by fedinskiy** [comment]
@michalvavrik just in case: this task is related to 3.8.6, and I am responsible for 3.15 :)

**Comment by michalvavrik** [comment]
> @michalvavrik just in case: this task is related to 3.8.6, and I am responsible for 3.15 :)

damn!

**Comment by michalvavrik** [comment]
In that case I'll leave this to @jcarranzan and @jedla97 and move to my tasks.

**Comment by jcarranzan** [comment]
Issue created: https://github.com/quarkusio/quarkus/issues/42854

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java
Comment (jedla97) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/YamlProvider.java:46]: I don't have problem with removal, just I don't see this class (`YamlProvider`) used anywhere in this PR and `@QuarkusApplication` specify the classes and this class is not included. So wondering if this shouldn't be in separate commit.

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:112]: As you created bug shoudn't we just check if  `.body(is("Headers response: ok"))` and disable the test with the link to your upstream issue. 

If it's really bug and will be fixed, this will break in daily run.

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersResource.java
Comment (jedla97) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersResource.java:38]: Afaik wouldn't be enough to send return just `Uni.createFrom().item("ok headers");` ? 
I don't say this is wrong! Just in regard to other endpoints only `headersOverride` using it and in addition to string  it's setting header.

**Review by jedla97 — CHANGES_REQUESTED** [review]
@jcarranzan Thanks for update. In overall it's look good to me. Just have 2 suggestion and also unresolve the thread about that logging with question.

**Comment by jedla97** [comment]
Also looking at the test and it seems that `HeadersMessageBodyWriter` changing all other tests so they are failing.

**Comment by jcarranzan** [comment]
> Also looking at the test and it seems that `HeadersMessageBodyWriter` changing all other tests so they are failing.

Fixed!, thanks @jedla97 .

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jcarranzan) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: Yes, u are right.

**Review by jedla97 — CHANGES_REQUESTED** [review]
Hi @jcarranzan I'm looking at this and don't like that this message header writer affecting other test and resources.

I looked at it and proposing solution.

That `MessageBodyWriter<String>` won't be catching a modify the all `String` responses but you will create new class (your own datatype) and it will be modified only the response types of this class. So the implementation will be something like `MessageBodyWriter<MySimpleClass>`. 

What I have looked it's valid usecase + you will have different test then is covered in upstream as bonus. I'll leave the implementation details on you. Just FYI I tried it with simple class and it worked.

Also when we use this implementation we don't need separate test class just for this one test (save few minutes when executing native :smile:)

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersResource.java
Comment (jedla97) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/HeadersResource.java:36]: Thinks that here you should return `Uni<CustomHeaderResponse>` ti intercept and modify the response

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java:101]: Can you revert this change as it's not part of the test coverage.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/DevModeGrpcIntegrationReactiveIT.java
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/DevModeGrpcIntegrationReactiveIT.java:73]: Can you revert this change as it's not part of the test coverage.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/HeadersIT.java:107]: Should be reactive test disabled?

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/BaseHttpAdvancedIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/BaseHttpAdvancedIT.java:114]: Can you revert this change as it's not part of the test coverage.

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:71]: Can you revert this change as it's not part of the test coverage.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersResource.java
Comment (jedla97) [inline http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/HeadersResource.java:22]: It would be nice to have same return String here as it's in reactive counter-part. You have there `"ok headers"`

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (jedla97) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:101]: As this is non reactive it should be probably disabled. If this work you should probably comment on that issue some findings

**Review by jedla97 — DISMISSED** [review]
Thanks for looking at proposed solution! I have few small comments .

Also I see that you checked the `Backport` in Pr description. This point is used when you backorting something to different branch.

Looking at `HeadersIT` the it's run on openshift `OpenShiftHeadersIT` so the `This change requires execution against OCP (use run tests phrase in comment)` point should be checked to not merge this before we run ocp tests

**Comment by jedla97** [comment]
run tests

**Comment by jedla97** [comment]
@michalvavrik can you do a review as I was finished this PR.

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (michalvavrik) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:100]: I refuse to believe that PR only changing independent RESTEasy Reactive library has to be tested in the RESTEasy Classic. They are completely independent implementations.

If you mean to test that same issue is not present in the RESTEasy Classic, I can respect that (though I would not), but I would not link the PR. Maybe link the issue and add comment that this is not reproducing, just verifying it is not there?

Comment (jedla97) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: TBH I was thinking yesterday if I should remove this classic coverage. Your reasoning make sense so I remove it.

@jcarranzan if you still think that we should test even the rest classic please create PR

Comment (michalvavrik) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: it's nice you be agreed with @jedla97 , but @rsvoboda doesn't :-) maybe just tweak the code.

my thinking was that users shouldn't use RESTEasy Classic anyway, so I care way less about it.

Comment (jedla97) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: Oh now I read @rsvoboda comment so I will update it back

Comment (jedla97) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: Reverted back + changed to use original issue in comment

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java
Comment (michalvavrik) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java:100]:
```suggestion
     * Coverage for https://github.com/quarkusio/quarkus/issues/41354 in RESTEasy Classic
```
optionally

Comment (jedla97) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/HeadersIT.java]: Done

**Review by michalvavrik — APPROVED** [review]

**Pull request merged**
